api/cluster/virtual_service.go

```diff
@@ -0,0 +1,228 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse model version revision url: %s", isvcURL)
+	}
+
+	if modelVersionRevisionURL.Scheme == "" {
+		veURL := "//" + isvcURL
+		modelVersionRevisionURL, err = url.Parse(veURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse model version revision url: %s", isvcURL)
+		}
+	}
+
+	return &VirtualService{
+		Name:                    fmt.Sprintf("%s-%s-%s", modelService.ModelName, modelService.ModelVersion, models.VirtualServiceComponentType),
+		Namespace:               modelService.Namespace,
+		ModelName:               modelService.ModelName,
+		VersionID:               modelService.ModelVersion,
+		RevisionID:              modelService.RevisionID,
+		Labels:                  modelService.Metadata.ToLabel(),
+		Protocol:                modelService.Protocol,
+		ModelVersionRevisionURL: modelVersionRevisionURL,
+	}, nil
+}
+
+func (cfg VirtualService) BuildVirtualServiceSpec() (*v1beta1.VirtualService, error) {
+	modelVersionHost, err := cfg.getModelVersionHost()
+	if err != nil {
+		return nil, err
+	}
+
+	modelVersionRevisionHost := cfg.ModelVersionRevisionURL.Hostname()
+	modelVersionRevisionPath := cfg.ModelVersionRevisionURL.Path
+
+	vs := &v1beta1.VirtualService{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "networking.istio.io/v1beta1",
+			Kind:       "VirtualService",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cfg.Name,
+			Namespace: cfg.Namespace,
+			Labels:    cfg.Labels,
+		},
+		Spec: istiov1beta1.VirtualService{
+			Gateways: []string{knativeIngressGateway},
+			Hosts:    []string{modelVersionHost},
+			Http:     cfg.createHttpRoutes(modelVersionRevisionHost, modelVersionRevisionPath),
+		},
+	}
+
+	return vs, nil
+}
+
+// getModelVersionHost creates model version endpoint host based on version endpoint's url
+func (cfg *VirtualService) getModelVersionHost() (string, error) {
+	host := strings.Split(cfg.ModelVersionRevisionURL.Hostname(), fmt.Sprintf(".%s.", cfg.Namespace))
+	if len(host) != 2 {
+		return "", fmt.Errorf("invalid version endpoint url: %s. failed to split domain: %+v", cfg.ModelVersionRevisionURL, host)
+	}
+
+	domain := host[1]
+	return fmt.Sprintf("%s-%s.%s.%s", cfg.ModelName, cfg.VersionID, cfg.Namespace, domain), nil
+}
+
+func (cfg *VirtualService) createHttpRoutes(modelVersionRevisionHost, modelVersionRevisionPath string) []*istiov1beta1.HTTPRoute {
+	routeDestinations := []*istiov1beta1.HTTPRouteDestination{
+		{
+			Destination: &istiov1beta1.Destination{
+				Host: defaultIstioIngressGatewayHost,
+			},
+			Headers: &istiov1beta1.Headers{
+				Request: &istiov1beta1.Headers_HeaderOperations{
+					Set: map[string]string{
+						"Host": modelVersionRevisionHost,
+					},
+				},
+			},
+			Weight: 100,
+		},
+	}
+
+	switch cfg.Protocol {
+	case protocol.UpiV1:
+		return []*istiov1beta1.HTTPRoute{
+			{
+				Route: routeDestinations,
+			},
+		}
+
+	default:
+		// Default to application/json if Content-Type header is empty
```

Comment:
Nit: I think having this comment here is a bit misleading - makes it look like we are adding json content-type header for all non-UPI protocol. But in fact, this comment is only relevant to the first route being returned. Shall we remove the comment here since there is a similar one below anyway ?